rust/yarp/src/lib.rs

```diff
@@ -0,0 +1,254 @@
+//! # yarp
+//!
+//! Rustified version of Ruby's YARP parser.
+//!
+#![warn(
+    box_pointers,
+    clippy::all,
+    clippy::nursery,
+    clippy::pedantic,
+    future_incompatible,
+    missing_docs,
+    nonstandard_style,
+    rust_2018_idioms,
+    trivial_casts,
+    trivial_numeric_casts,
+    unreachable_pub,
+    unused_qualifications
+)]
+
+// Most of the code in this file is generated, so sometimes it generates code
+// that doesn't follow the clippy rules. We don't want to see those warnings.
+#![allow(
+    clippy::too_many_lines,
+    clippy::use_self
+)]
+
+mod bindings {
+    // In `build.rs`, we generate bindings based on the config.yml file. Here is
+    // where we pull in those bindings and make them part of our library.
+    include!(concat!(env!("OUT_DIR"), "/bindings.rs"));
+}
+
+pub use self::bindings::*;
+
+use std::ffi::{CStr, CString};
+use std::mem::MaybeUninit;
+use yarp_sys::{yp_parser_t, yp_parser_init, yp_parse, yp_node_destroy, yp_node_t, yp_comment_t, yp_diagnostic_t};
+
+use std::ptr::NonNull;
+
+/// A diagnostic message that came back from the parser.
+#[derive(Copy, Clone, Debug)]
+pub struct Diagnostic(*mut yp_diagnostic_t);
+
+impl Diagnostic {
+    /// Returns the message associated with the diagnostic.
+    /// 
+    /// # Panics
+    /// 
+    /// Panics if the message is not able to be converted into a `CStr`.
+    /// 
+    #[must_use]
+    pub fn message(&self) -> String {
+        unsafe {
+            let message: *mut i8 = (*(self.0)).message as *mut i8;
+            CStr::from_ptr(message).to_str().unwrap().to_string()
+        }
+    }
+}
+
+/// A comment that was found during parsing.
+#[derive(Copy, Clone, Debug)]
+pub struct Comment(*mut yp_comment_t);
+
+impl Comment {
+    /// Returns the text of the comment.
+    #[must_use]
+    pub fn text(&self) -> String {
+        unsafe {
+            let start: *mut u8 = (*(self.0)).start as *mut u8;
+            let end: *mut u8 = (*(self.0)).end as *mut u8;
+
+            let length = end as usize - start as usize;
+            String::from_raw_parts(start, length, length)
```

Comment:
this is unsound. `String::from_raw_parts` can only be safely used on raw parts derived from a string.

Instead, create a `&[u8]`, call `std::str::from_utf8`, and then create an owned string.

However, given that the memory is already owned by the parser, it seems more acceptable to return a `&str` bound to the lifetime of the `Comment`.

One other question I have: are the byte contents guaranteed to be UTF-8 even of the source encoding of the file is e.g. shift JIS?

Based on the multi-encoding support in the parser I expected most of these APIs to be written in terms of bytes.

Reply:
The bytes in the comment message specifically are guaranteed to be UTF-8 regardless of the source because we're only using UTF-8.